packages/structures/src/bitfields/BitField.ts

```diff
@@ -0,0 +1,203 @@
+		return this.bitfield === this.constructor.resolve(bit);
+	}
+
+	/**
+	 * Checks whether the bitfield has a bit, or multiple bits.
+	 *
+	 * @param bit - Bit(s) to check for
+	 * @returns
+	 */
+	public has(bit: BitFieldResolvable<Flags>, ..._hasParams: unknown[]) {
+		const resolvedBit = this.constructor.resolve(bit);
+		return (this.bitfield & resolvedBit) === resolvedBit;
+	}
+
+	/**
+	 * Gets all given bits that are missing from the bitfield.
+	 *
+	 * @param bits - Bit(s) to check for
+	 * @param hasParams -  Additional parameters for the has method, if any
+	 * @returns
+	 */
+	public missing(bits: BitFieldResolvable<Flags>, ...hasParams: readonly unknown[]) {
+		return new (this.constructor as unknown as new (bits: BitFieldResolvable<Flags>) => BitField<Flags>)(bits)
+			.remove(this)
+			.toArray(...hasParams);
+	}
+
+	/**
+	 * Freezes these bits, making them immutable.
+	 *
+	 * @returns
+	 */
+	public freeze() {
+		return Object.freeze(this);
+	}
+
+	/**
+	 * Adds bits to these ones.
+	 *
+	 * @param bits - Bits to add
+	 * @returns These bits or new BitField if the instance is frozen.
+	 */
+	public add(...bits: BitFieldResolvable<Flags>[]) {
+		let total = this.constructor.DefaultBit;
+		for (const bit of bits) {
+			total |= this.constructor.resolve(bit);
+		}
+
+		if (Object.isFrozen(this)) return new this.constructor(this.bitfield | total);
+		this.bitfield |= total;
+		return this;
+	}
+
+	/**
+	 * Removes bits from these.
+	 *
+	 * @param bits - Bits to remove
+	 * @returns These bits or new BitField if the instance is frozen.
+	 */
+	public remove(...bits: BitFieldResolvable<Flags>[]) {
+		let total = this.constructor.DefaultBit;
+		for (const bit of bits) {
+			total |= this.constructor.resolve(bit);
+		}
+
+		if (Object.isFrozen(this)) return new this.constructor(this.bitfield & ~total);
+		this.bitfield &= ~total;
+		return this;
+	}
+
+	/**
+	 * Gets an object mapping field names to a {@link boolean} indicating whether the
+	 * bit is available.
+	 *
+	 * @param hasParams - Additional parameters for the has method, if any
+	 * @returns
+	 */
+	public serialize(...hasParams: readonly unknown[]) {
+		const serialized: Partial<Record<keyof Flags, boolean>> = {};
+		for (const [flag, bit] of Object.entries(this.constructor.Flags)) {
+			if (Number.isNaN(Number(flag))) serialized[flag as keyof Flags] = this.has(bit as bigint | number, ...hasParams);
+		}
+
+		return serialized;
+	}
+
+	/**
+	 * Gets an {@link Array} of bitfield names based on the bits available.
+	 *
+	 * @param hasParams - Additional parameters for the has method, if any
+	 * @returns
+	 */
+	public toArray(...hasParams: readonly unknown[]) {
+		return [...this[Symbol.iterator](...hasParams)];
+	}
+
+	public toJSON(asNumber?: boolean) {
+		if (asNumber) {
+			if (this.bitfield > Number.MAX_SAFE_INTEGER) {
+				throw new RangeError(
+					`Cannot convert bitfield value ${this.bitfield} to number, as it is bigger than Number.MAX_SAFE_INTEGER`,
```

Comment:
Might as well

```suggestion
					`Cannot convert bitfield value ${this.bitfield} to number, as it is bigger than ${Number.MAX_SAFE_INTEGER}`,
```

Reply:
For an uninformed user that will just look like a "random" number though...
While as it is currently, it shows them where the value comes from.